### ISSUE TITLE
Pathfinder can jump/fall through stone

### ISSUE THREAD

**Issue by MarkuBu** [minetest/minetest#4198]
I set jump/fall height to 10 and I was wondering why the pathfinder can find me in a cave, where the start point of the path is above the surface and the end is in the cave. 

The pathfinder can jump/fall through solid nodes

**Comment by est31** [comment]
Yeah this max jump/fall stuff is totally broken, and in fact a lie. its not about falling/jumping, but about getting through solid matter. At least that's what the code says.

**Comment by MarkuBu** [comment]
This is a mess. And I think sometimes the pathfinder stucks and blocks the game

**Comment by est31** [comment]
There has been a big performance problem especially with longer distances, but if you use 0.4.14 or later it should be faster.

Still, the pathfinder has many other problems and @gregorycu is working on a replacement.

**Comment by MarkuBu** [comment]
Good news. Then I don't need to write my own

**Comment by gregorycu** [comment]
Sorry for the slow process on this. I'll work on the pathfinder this
weekend. I just need to clean up some code.
On 08/06/2016 7:25 am, "MarkuBu" <notifications@github.com> wrote:

Good news. Then I don't need to write my own

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/minetest/minetest/issues/4198#issuecomment-224419295>,
or mute the thread
<https://github.com/notifications/unsubscribe/ABcig6ESxoqdbNBIvAcTxf2PNLCxIENXks5qJeHGgaJpZM4IwA88>
.

**Comment by MarkuBu** [comment]
Has the pathfinder new features (flying, swimming) or is it just a faster and less buggy version of the current?

**Comment by gregorycu** [comment]
The new pathfinder can find a 20 node path in about 1ms. I think it's faster than the old one.

It's a complete rewrite, so if it has bugs, it probably has different bugs to the old one, though I am not aware of any bugs at the moment.

With regards to swimming, it support swimming on-top of the water, like a duck, swimming inside the water but only near the surface, like a dog, and under the surface, like a fish.

An example video is here: https://www.youtube.com/watch?v=8vFKXovblCw

I have not implemented flying yet, but it shouldn't be too hard to add.

**Comment by MarkuBu** [comment]
Looks great. Did you also add climbing (e.g. spider)?

**Comment by gregorycu** [comment]
No, I didn't. In that video, the cat is jumping.

I can add climbing. Probably harder than flying, so I'll do flying first.

More than happy for other suggestions too.

**Comment by MarkuBu** [comment]
I guess you want to be compatible to the current code. I mean, you just return a list of coordinates of the path. But it would be nice if you could add an optional yaw to the next node. 

For example, if the cat in the video jumps up the nodes she had to turn around. 

e.g.
```
{
 pos={x=0, y=0, z=0},
  yaw=0
}
```

Maybe other informations will be helpful like a jump value. 0 means no jump to the next node, positive mean jump (height), negative values mean fall (height)

```
{
  pos={x=0, y=0, z=0},
  yaw=0, 
  jump=2
}
```

So any information the pathfinder has while he is running and might be useful for e.g. the mob api can be added to the path. This might be slow the pathfinder a bit, but helps the following steps. The mob api can set a jump height without calculation. It looks odd if a mob jumps 4 nodes high to get up one node. Or the api can calculate fall damage.

To speed up you can use multitasking, if you don't use it already. Pathfinder algorithms are perfect for multithasking. Are you using A* or HPA*?

**Comment by gregorycu** [comment]
Just A* for me.

Yes, there is a lot of information that could be returned. For instance,
how to get to the position in question.

In the video of the cat, yeah, it was a little weird that she took that
route, what the pathfinder needs is certain weightings for certain actions.
A cat jumping up may cost 1.3 movement points.

I could certainly look at how the algo handles directions, to be honest, it
may require a bit of an architectural change to the algo.

**Comment by MarkuBu** [comment]
Another idea is to add a walkabel width and heigh, for example a normal mob has a height of 2 and a width of 1 and can walk through a door. A gobblin can walk through a one block height, a spider needs a width of 2

The current mobs are dopy. I stand in a shed with a door and a window and the mobs can't find a way through the door. They only jump around outside.

Problem with a width of 2 would be, that the center of a mob is between the two nodes.

**Comment by gregorycu** [comment]
I already handle heights of more than 1. Widths are a different ball-game
(as you identified)

**Comment by MarkuBu** [comment]
Is it possible to add the height as parameter?

Width would be important because if mobs like spiders walk along a wall they have collisions or parts of the object are inside the wall. Maybe you can add an offset x+0.5, z+0.5 to move these mobs from the center of an node to the corner and use 2x2 nodes instead of one node when search the path

**Comment by MarkuBu** [comment]
Another thing I saw in this video
https://youtu.be/KNXfSOx4eEE?t=725

You should avoid diagonal path along an obstacle. If you return a path like this, mobs would run into the obstacle and would "slide" along until they passed it

**Comment by MarkuBu** [comment]
@gregorycu any progress on the pathfinder replacement?

**Comment by Wuzzy2** [comment]
Fixed by #9339.
